core/trino-main/src/main/java/io/trino/operator/exchange/ScaleWriterPartitioningExchanger.java

```diff
@@ -0,0 +1,196 @@
+            if (partitionSize == 0) {
+                continue;
+            }
+
+            // clear the assigned positions list size for the next iteration to start empty. This
+            // only resets the size() to 0 which controls the index where subsequent calls to add()
+            // will store new values, but does not modify the positions array
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            int nextPartitionWriterId = getNextWriterId(partitionId);
+            pagePartitionsRowCount.put(serialize(new WriterPartitionId(nextPartitionWriterId, partitionId)), partitionSize);
+
+            for (int i = 0; i < partitionSize; i++) {
+                writerAssignments[nextPartitionWriterId].add(positions[i]);
+            }
+        }
+
+        // Update the partitions row count state which will help with scaling partitions across writers
+        updatePartitionRowCount(pagePartitionsRowCount);
+
+        // build a page for each writer
+        for (int bucket = 0; bucket < writerAssignments.length; bucket++) {
+            IntArrayList positionsList = writerAssignments[bucket];
+            int bucketSize = positionsList.size();
+            if (bucketSize == 0) {
+                continue;
+            }
+            // clear the assigned positions list size for the next iteration to start empty. This
+            // only resets the size() to 0 which controls the index where subsequent calls to add()
+            // will store new values, but does not modify the positions array
+            int[] positions = positionsList.elements();
+            positionsList.clear();
+
+            if (bucketSize == page.getPositionCount()) {
+                // whole input page will go to this partition, compact the input page avoid over-retaining memory and to
+                // match the behavior of sub-partitioned pages that copy positions out
+                page.compact();
+                sendPageToPartition(buffers.get(bucket), page);
+                return;
+            }
+
+            Page pageSplit = page.copyPositions(positions, 0, bucketSize);
+            sendPageToPartition(buffers.get(bucket), pageSplit);
+        }
+    }
+
+    private synchronized void updatePartitionRowCount(Long2IntMap pagePartitionsRowCount)
+    {
+        pagePartitionsRowCount.forEach((writerPartitionId, rowCount) ->
+                partitionsRowCount.merge(writerPartitionId, rowCount, Long::sum));
+    }
+
+    public synchronized Long2LongMap getAndResetPartitionsRowCount()
+    {
+        Long2LongMap result = new Long2LongOpenHashMap(partitionsRowCount);
+        partitionsRowCount.clear();
+        return result;
+    }
+
+    private int getNextWriterId(int partitionId)
+    {
+        int currentIndex = partitionWriterIndex[partitionId];
+        partitionWriterIndex[partitionId] = (currentIndex + 1) % partitionRebalancer.getWriterCount(partitionId);
```

Comment:
`partitionRebalancer.getWriterCount(partitionId);` method is not needed.
You can just apply modulo ([Math.floorMod()](https://docs.oracle.com/javase/8/docs/api/java/lang/Math.html#floorMod-int-int-)) inside of `getWriterId`

Reply:
In that case, don't we have to manage current indexes inside `UniformPartitionRebalancer`? And that'll be inefficient since we have to use some concurrent data structure.

Reply:
What I mean is that you can do:
```
partitionRebalancer.getWriterId(partitionId, currentIndex);
```
and
`getWriterId` could apply `currentIndex % partitionRebalancer.getWriterCount(partitionId);` internally